File: src/api/routers/standalone_dub.py
"""Standalone SRT → Dub CRUD routes."""

from __future__ import annotations

import asyncio

from fastapi import APIRouter, File, Form, HTTPException, UploadFile
from starlette.responses import FileResponse

from src.api import standalone_dub as standalone_mod
from src.api.deps import get_config, get_task_manager
from src.api.models import TaskResponse

router = APIRouter()


@router.post(
    "/api/standalone-dub",
    response_model=TaskResponse,
    status_code=201,
)
async def start_standalone_dub(
    file: UploadFile = File(...),
    provider: str = Form(...),
    voice: str = Form(...),
    language: str = Form(...),
    playback_speed: float = Form(1.5),
    enable_shortening: bool = Form(True),
    api_key: str | None = Form(None),
    llm_api_key: str | None = Form(None),
    llm_backend: str | None = Form(None),
):
    """Generate a dub from an uploaded SRT. Returns task_id; subscribe
    to the existing /api/tasks/{task_id} SSE for progress."""
    tm = get_task_manager()
    config = get_config()

    content = await file.read()
    task = tm.create_task("standalone_dub")
    task._asyncio_task = asyncio.create_task(
        tm.run_standalone_dub(
            task_id=task.task_id,
            srt_content=content,
            original_filename=file.filename or "uploaded.srt",
            provider=provider,
            voice=voice,
            language=language,
            playback_speed=playback_speed,
            enable_shortening=enable_shortening,
            config=config,
            api_key_override=api_key,
            llm_api_key=llm_api_key,
            llm_backend=llm_backend,
        )
    )
    return TaskResponse(task_id=task.task_id, status=task.status)


@router.get(
    "/api/standalone-dub",
    response_model=list[standalone_mod.StandaloneDubEntry],
)
async def list_standalone_dubs():
    """Recent dubs, newest first."""
    return standalone_mod.list_dubs()


@router.delete("/api/standalone-dub/{dub_uuid}", status_code=204)
async def delete_standalone_dub(dub_uuid: str):
    """Remove the WAV + metadata sidecar."""
    ok = standalone_mod.delete_dub(dub_uuid)
    if not ok:
        raise HTTPException(status_code=404, detail=f"Dub {dub_uuid} not found")
    return None


@router.get("/api/standalone-dub/{dub_uuid}.wav")
async def download_standalone_dub(dub_uuid: str):
    """Serve the WAV with Content-Disposition: attachment for download."""
    wav = standalone_mod.wav_path(dub_uuid)
    if not wav.exists():
        raise HTTPException(status_code=404, detail=f"Dub {dub_uuid} not found")
    return FileResponse(
        path=str(wav),
        media_type="audio/wav",
        filename=f"{dub_uuid}.wav",
    )

File: ui-app/src/api/standaloneDub.ts
import type { TaskResponse } from './types';

export interface StandaloneDubEntry {
  uuid: string;
  original_filename: string;
  provider: string;
  voice: string;
  language: string;
  playback_speed: number;
  enable_shortening: boolean;
  duration_seconds: number;
  created_at: string;
  file_size_bytes: number;
}

export interface PostStandaloneDubOpts {
  file: File;
  provider: string;
  voice: string;
  language: string;
  playbackSpeed: number;
  enableShortening: boolean;
  apiKey?: string;
  llmApiKey?: string;
  llmBackend?: string;
  model?: string;
}

export async function postStandaloneDub(
  opts: PostStandaloneDubOpts,
): Promise<TaskResponse> {
  const formData = new FormData();
  formData.append('file', opts.file);
  formData.append('provider', opts.provider);
  formData.append('voice', opts.voice);
  formData.append('language', opts.language);
  formData.append('playback_speed', String(opts.playbackSpeed));
  formData.append('enable_shortening', String(opts.enableShortening));
  if (opts.apiKey) formData.append('api_key', opts.apiKey);
  if (opts.llmApiKey) formData.append('llm_api_key', opts.llmApiKey);
  if (opts.llmBackend) formData.append('llm_backend', opts.llmBackend);
  if (opts.model) formData.append('model', opts.model);

  const r = await fetch('/api/standalone-dub', {
    method: 'POST',
    body: formData,
  });
  if (!r.ok) {
    const body = await r.text().catch(() => '');
    throw new Error(`${r.status} ${body}`);
  }
  return r.json();
}

export async function getStandaloneDubs(): Promise<StandaloneDubEntry[]> {
  const r = await fetch('/api/standalone-dub');
  if (!r.ok) {
    throw new Error(`${r.status}`);
  }
  return r.json();
}

export async function deleteStandaloneDub(dubUuid: string): Promise<void> {
  const r = await fetch(`/api/standalone-dub/${dubUuid}`, { method: 'DELETE' });
  if (!r.ok && r.status !== 204) {
    throw new Error(`${r.status}`);
  }
}

export function getStandaloneDubUrl(dubUuid: string): string {
  return `/api/standalone-dub/${dubUuid}.wav`;
}
